pkg/controller/endpointslice/endpointslice_tracker.go

```diff
@@ -19,102 +19,147 @@ package endpointslice
 import (
 	"sync"
 
+	"k8s.io/api/core/v1"
 	discovery "k8s.io/api/discovery/v1beta1"
 	"k8s.io/apimachinery/pkg/types"
 )
 
-// endpointSliceResourceVersions tracks expected EndpointSlice resource versions
-// by EndpointSlice name.
-type endpointSliceResourceVersions map[string]string
+const (
+	deletionExpected = -1
+)
+
+// generationsBySlice tracks expected EndpointSlice generations by EndpointSlice
+// uid. A value of deletionExpected (-1) may be used here to indicate that we
```

Comment:
just naive question for @wojtek-t , the generation field is not something that can be crafted and modified by a user to set this value and cause possible problems, right?

Reply:
It can to some extent - here you have the logic:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/discovery/endpointslice/strategy.go#L57

If someone will try to change that without changing the "real contents" of EPS, we won't prevent it.
It's pretty much the logic we use elsewhere though:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/apps/deployment/strategy.go#L106


But we may consider setting it to the same value if the spec didn't change (though if so, let's do that in a separate PR - I would like someone like Jordan to take a quick look at it).

Reply:
This is a good idea, thanks! I'll follow a separate PR for this. Do you happen to know if changes to strategy can ever be backported? I'm interested in cherrypicking this back to 1.20 or 1.19 if possible, but I realize the scope of this change is already pretty large.

Reply:
That's a good question - the fact that they are in strategy doesn't mean they can't be backported, but this would be in fact strenghtening validation, so probably not.